infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -0,0 +1,225 @@
+  std::vector<std::byte> memory;
+
+  for (size_t i = 0; i < size; i += sizeof(long)) {
+    long word = ptrace(PTRACE_PEEKTEXT, pid, address + i, 0);
+    if (word == -1) {
+      return memory;
+    }
+
+    std::byte *word_bytes = reinterpret_cast<std::byte*>(&word);
+    memory.insert(memory.end(), word_bytes, word_bytes+sizeof(long));
+  }
+
+  return memory;
+}
+
+void inspect(pid_t pid, const user_regs_struct &regs) {
+  auto memory = read_memory(pid, regs.rdi, kTripWire.length());
+  if (memory.size() == 0) {
+    return;
+  }
+
+  std::string path(reinterpret_cast<char*>(
+        memory.data()), std::min(memory.size(), kTripWire.length()));
+  debug_log("inspecting");
+  if (path == kTripWire) {
+    kill(pid, SIGKILL);
+    printf("===BUG DETECTED: Shell injection===\n");
+    // TODO: Get/print stacktrace.
+    exit(1);
+  }
+}
+
+void trace(std::set<pid_t> pids) {
+  while (!pids.empty()) {
+    std::vector<pid_t> new_pids;
+
+    auto it = pids.begin();
+    while (it != pids.end()) {
+      auto pid = *it;
+      int status = 0;
+
+      int result = waitpid(pid, &status, __WALL | WNOHANG);
+      if (result == -1) {
+        it = pids.erase(it);
+        continue;
+      }
+
+      if (result == 0) {
+        // Nothing to report yet.
+        ++it;
+        continue;
+      }
+
+      debug_log("finished waiting %d", pid);
+
+      if (WIFEXITED(status) || WIFSIGNALED(status)) {
+        debug_log("%d exited", pid); 
+        it = pids.erase(it);
+        continue;
+      }
+
+      bool is_syscall = WIFSTOPPED(status) && WSTOPSIG(status) == (SIGTRAP | 0x80);
+      int sig = 0;
+      if (!is_syscall) {
+        // Handle generic signal.
+        siginfo_t siginfo;
+        if (ptrace(PTRACE_GETSIGINFO, pid, nullptr, &siginfo) == -1) {
+          debug_log("ptrace(PTRACE_GETSIGINFO, %d): %s", pid, strerror(errno));
+          continue;
+        }
+        sig = siginfo.si_signo;
+        debug_log("forwarding signal %d to %d", sig, pid);
+      }
+
+      if (WIFSTOPPED(status) && 
+          (status>>8 == (SIGTRAP | (PTRACE_EVENT_CLONE<<8)) ||
+           status>>8 == (SIGTRAP | (PTRACE_EVENT_FORK<<8)) ||
+           status>>8 == (SIGTRAP | (PTRACE_EVENT_VFORK<<8)))) {
+        long new_pid;
+        if (ptrace(PTRACE_GETEVENTMSG, pid, 0, &new_pid) == -1) {
+          debug_log("ptrace(PTRACE_GETEVENTMSG, %d): %s", pid, strerror(errno));
+          continue;
+        }
+        debug_log("forked %ld", new_pid);
+        new_pids.push_back(new_pid);
+      }
+
+      if (is_syscall) {
+        // TODO: distinguish between syscall enter and exit.
+        user_regs_struct regs;
+        if (ptrace(PTRACE_GETREGS, pid, 0, &regs) == -1) {
+          debug_log("ptrace(PTRACE_GETREGS, %d): %s", pid, strerror(errno));
+          continue;
+        }
+
+        if (regs.orig_rax == __NR_execve) {
+          inspect(pid, regs);
+        }
+        // TODO: Check for commands with invalid syntax passed to /bin/sh and
```

Comment:
I realize that checking for proof tripwire was executed can catch unlikely edge cases like if our process communicates with another one that then executes tripwire that this misses.

Reply:
that's true. i'll leave a TODO here for that case.

The challenge with trying to catch actual execution is race conditions wrt these syscalls (since execve doesn't block), and not everyone will call waitpid properly. Maybe that can be a more lightweight check where we only check for this after the entire program finishes running.